enterprise/metrics/src/main/java/org/neo4j/metrics/source/JvmMetrics.java

```diff
@@ -0,0 +1,220 @@
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.lang.management.BufferPoolMXBean;
+import java.lang.management.GarbageCollectorMXBean;
+import java.lang.management.ManagementFactory;
+import java.lang.management.MemoryPoolMXBean;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.management.ListenerNotFoundException;
+import javax.management.Notification;
+import javax.management.NotificationEmitter;
+import javax.management.NotificationListener;
+import javax.management.openmbean.CompositeData;
+
+import org.neo4j.kernel.configuration.Config;
+import org.neo4j.metrics.MetricsSettings;
+
+import static com.codahale.metrics.MetricRegistry.name;
+import static com.sun.management.GarbageCollectionNotificationInfo.GARBAGE_COLLECTION_NOTIFICATION;
+
+public class JvmMetrics implements Closeable
+{
+    private static final String NAME_PREFIX = "neo4j.vm";
+    private static final String GC = name( NAME_PREFIX, "gc" );
+    private static final String MEMORY_POOL = name( NAME_PREFIX, "memory.pool" );
+    private static final String MEMORY_BUFFER = name( NAME_PREFIX, "memory.buffer" );
+    private static final String THREAD = name( NAME_PREFIX, "thread" );
+    private final Config config;
+    private final MetricRegistry registry;
+    private final List<Runnable> removeListenerHandlers = new ArrayList<>();
+
+    public JvmMetrics( Config config, MetricRegistry registry )
+    {
+        this.config = config;
+        this.registry = registry;
+
+        // VM stats
+        // Garbage collection
+        if ( config.get( MetricsSettings.jvmGcEnabled ) )
+        {
+            for ( GarbageCollectorMXBean gcBean : ManagementFactory.getGarbageCollectorMXBeans() )
+            {
+                final AtomicLong periodGcDuration = new AtomicLong();
+                registry.register( name( GC, prettifyName( gcBean.getName() ) ), new Gauge<Long>()
+                {
+                    public Long getValue()
+                    {
+                        return periodGcDuration.get();
+                    }
+                } );
+
+                final NotificationListener listener = new NotificationListener()
+                {
+                    public void handleNotification( Notification notification, Object handback )
+                    {
+                        if ( notification.getType().equals( GARBAGE_COLLECTION_NOTIFICATION ) )
+                        {
+                            GarbageCollectionNotificationInfo info = GarbageCollectionNotificationInfo.from(
+                                    (CompositeData) notification.getUserData() );
+                            long duration = info.getGcInfo().getDuration();
+                            periodGcDuration.addAndGet( duration );
+                        }
+                    }
+                };
+
+                final NotificationEmitter emitter = (NotificationEmitter) gcBean;
```

Comment:
I think we guard this cast (or do an early `continue` at the top of the loop body) with an `instanceof` check, because it is making assumptions that can be JVM vendor specific.